bindings/python/slimt.cpp

```diff
@@ -142,6 +144,103 @@ class PyService {
+    // Ranges are in C++ Byte Ranges.  Python however requires indices into
+    // unicode strings. For example, the following is required without this
+    // change.
+    //
+    // # Case 1: Byte Ranges
+    //
+    // source = response.source.text // "Hello world"
+    // range = response.source.word_as_range(0, 0)
+    //
+    // #  Additional encoding step required to access
+    // #  "byte" ranges.
+    // bytes = source.encode("utf8")
+    //
+    // word_bytes = bytes[range.begin:range.end]
+    // word = word_bytes.decode("utf8")
+    //
+    // # Case 2: Unicode (UTF-8 ranges)
+    //
+    // What we want to do, is omit the conversion step, by providing ranges
+    // native to python.
+    //
+    // source = response.source.text // "Hello world"
+    // range = response.source.word_as_range(0, 0)
+    // word = source[range.begin:range.end]
+    //
+    //
+    // Assert:
+    //  word == response.source.word(0, 0)
+    //  [python-slicing] = C++ string conversion
+    //
+    // Why? Diagnosis?
+    //
+    // pybind11 converts std::string, by default to unicode-strings (python).
+    // However, since we manage ranges, they remain "byte ranges". We can
+    // mitigate this problem, by patching the annotation's ranges to point to
+    // unicode before returning the response.
+
+    auto transform_to_utf8 =
+        [](const AnnotatedText &annotated) -> std::vector<Range> {
+      // f: Range [Bytes] -> Range[UTF8]
+      const std::string &text = annotated.text;
+
+      size_t sentence_idx = 0;
+      size_t word_idx = 0;
+      Range current = annotated.word_as_range(sentence_idx, word_idx);
+
+      std::vector<Range> words;
+
+      size_t utf8_idx = 0;
+      size_t byte_idx = 0;
+      Range utf8{
+          //
+          .begin = utf8_idx,  //
+          .end = 0            //
+      };
+      // This loops run on the entire string.
+      //
+      // We have indices into two views of the same string. One is bytes, the
+      // other is utf8 encoded. We want to convert what is bytes to utf8.
+      //
+      // For this, we traverse through the characters, checking for unicode
+      // encoding and associated adjustments to the utf8_idx, keeping
+      // correspondences with the byte_idx;
+      for (const char c : text) {
+        // current = [begin, end)
+        if ((c & 0xc0) != 0x80)  // if is not utf-8 continuation character
+          ++utf8_idx;
+
+        ++byte_idx;
+
+        if (byte_idx == current.end) {
+          utf8.end = utf8_idx;
+
+          // Push it to the list of (utf8) words.
+          // We will use these to create the utf8 range annotation.
+          words.push_back(utf8);
+
+          ++word_idx;
+          if (word_idx >= annotated.word_count(sentence_idx)) {
+            ++sentence_idx;
+            word_idx = 0;
+          }
+
+          current = annotated.word_as_range(sentence_idx, word_idx);
+
+          utf8.begin = utf8_idx;
+          utf8.end = -1;
+        }
+      }
```

Comment:
There's some code here that's _traverses through words_. One time-tested suitable mechanism to do this is an [`iterator`](https://en.cppreference.com/w/cpp/iterator/iterator). 

```cpp
class WordIterator {
  public:
    WordIterator(const AnnotatedText &annotated): annotated_(annotated){}
    // Implement ++, --, etc. We can keep this minimum and enhance as required, FWIW
    // No need to follow full iterator spec.
  private:
    const AnnotatedText &annotated_;
    size_t sentence_id_;
    size_t word_id_; 
};
```

This will lead to a calling code:

```cpp
   WordIterator current(annotated);
   // ... 
   
   for(char c: annotated.text){
     // ...
     current++;
   }
```

Hiding away some stuff that makes this section _unpleasant_.

Reply:
There are plans of reusing `WordIterator`. We're trying to make the Python API pleasant using something similar:

<details> <summary> Python: Word, Sentence Iterators </summary>

```py
# Wrappers to hide ugliness, easy prototyping in python.
# Eventually we'll push this down to pybind11/C++
class WordIterator:
  def __init__(self, annotation, sentence_id):
    self._sentence_id = sentence_id
    self._annotation = annotation

  def __iter__(self):
    self._word_id = -1
    return self

  def __next__(self):
    self._word_id += 1
    if self._word_id >= self._annotation.word_count(self._sentence_id):
      raise StopIteration
    return self

  def surface(self):
    range = self.range()
    return self._annotation.text[range.begin:range.end]

  def range(self):
    return self._annotation.word_as_range(self._sentence_id, self._word_id)

  def id(self):
    return (self._sentence_id, self._word_id)

class SentenceIterator:
  def __init__(self, annotation):
    self._annotation = annotation
    self._sentence_id = -1

  def __iter__(self):
    self._sentence_id = -1
    return self

  def __next__(self):
    self._sentence_id += 1
    if self._sentence_id >= self._annotation.sentence_count():
      raise StopIteration
    return self

  def words(self):
    return WordIterator(self._annotation, self._sentence_id)

  def __repr__(self):
    return f'{self._sentence_id}'
```
</details>